[clang][dataflow] Make optional checker work for types derived from optional.

martinboehme opened the pull request:
`llvm::MaybeAlign` does this, for example.

It's not an option to simply ignore these derived classes because they get cast
back to the optional classes (for example, simply when calling the optional
member functions), and our transfer functions will then run on those optional
classes and therefore require them to be properly initialized.

ymand commented:
Per offline discussion, please see if we can be more targeted in our use of the generalized matching, given the potentially high cost incurred (for every relevant operation, like comparison, we'll be searching the entire type hierchy of the arguments).

martinboehme committed:
[clang][dataflow] Make optional checker work for types derived from optional.

`llvm::MaybeAlign` does this, for example.

It's not an option to simply ignore these derived classes because they get cast
back to the optional classes (for example, simply when calling the optional
member functions), and our transfer functions will then run on those optional
classes and therefore require them to be properly initialized.

martinboehme commented on clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp:
```diff
@@ -64,38 +64,53 @@ static bool hasOptionalClassName(const CXXRecordDecl &RD) {
+    return nullptr;
+
+  for (const CXXBaseSpecifier &Base : RD->bases())
+    if (Base.getAccessSpecifier() == AS_public)
```
Great point! I've removed the `AS_public` check (and added a test for exactly this scenario, which fails with this check present).

ymand commented on clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp:
```diff
@@ -64,39 +64,117 @@ static bool hasOptionalClassName(const CXXRecordDecl &RD) {
+      break;
+    Ty = Base->getType();
+    TyIsThisType = false;
+  }
```
Why not break here? (that is, only ever examine at most one `Base`)? Is the idea that we want the last in the list of base classes? Might be worth a comment.

martinboehme replied:
> Is the idea that we want the last in the list of base classes?

Exactly. I've added comments that will hopefully make it clearer what is going on here.

ymand replied:
Yes, very clear now. Thanks!

ymand commented on clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp:
```diff
@@ -64,39 +64,117 @@ static bool hasOptionalClassName(const CXXRecordDecl &RD) {
+  // base class even if the base is non-public.
+  bool TyIsThisType = isa<CXXThisExpr>(Cast->getSubExpr());
+
+  for (const CXXBaseSpecifier *Base : Cast->path()) {
```
what does `path` do? Sadly, the clang header file had no comments...

martinboehme replied:
I've submitted https://github.com/llvm/llvm-project/pull/85623 to add documentation.

ymand commented on clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp:
```diff
@@ -64,39 +64,117 @@ static bool hasOptionalClassName(const CXXRecordDecl &RD) {
-              anyOf(hasOptionalType(),
-                    hasType(pointerType(pointee(optionalOrAliasType())))))),
-      callee(cxxMethodDecl(matcher)));
+  return cxxMemberCallExpr(Ignorable ? on(expr(unless(*Ignorable)))
```
why drop the exclusion of calls on `this`? is it that this scenario is no longer possible given the more precise matching logic?

martinboehme replied:
> why drop the exclusion of calls on `this`?

We do this so that we can analyze member functions in classes that are derived from an optional class. See the new test `ClassDerivedPrivatelyFromOptional` (added in response to @Xazax-hun 's earlier [comment](https://github.com/llvm/llvm-project/pull/84138#discussion_r1514849752)) , which obviously fails (because it doesn't get analyzed) if we disallow `cxxThisExpr` here.

The `cxxThisExpr` exclusion was present in the [patch](https://reviews.llvm.org/D121197) that originally added this check. The exclusion wasn't discussed at the time, but I assume the intent was to avoid analyzing member functions of the optional types themselves. I would assume the savings provided by this are relatively small, and this is offset by a) the additional complexity in the AST pattern, and b) the fact that it prevents us from analyzing code such as `ClassDerivedPrivatelyFromOptional`. (I realize that deriving privately from an optional type is likely rare in practice, and probably a bad idea, but if we can support it by removing complexity, that seems like a good tradeoff.)

ymand commented on clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp:
```diff
@@ -64,39 +64,117 @@ static bool hasOptionalClassName(const CXXRecordDecl &RD) {
+    return Ty;
+  }
+
+  QualType Ty = getPublicType(Cast->getSubExpr());
```
I don't understand this line. We're recursing, but then we throw away the result `Ty` based on how the loop turns out. Why not just test for this directly, after the loop and only recurse when the loop doesn't find what we want?  See more comments below, but perhaps the problem is that I don't really understand what the loop is doing.

martinboehme replied:
I hope this has become clearer given the additional comments on what the loop is doing.

The idea behind doing the recursion first, then potentially throwing the result away, is that it makes the code proceed strictly along the patch from most-derived to most-base class.

Conceptually, we want to "drill down" into the `getImplicitObjectArgument()` to find the most-derived type, then walk back along the "cast path" towards the base until we find a point where we're no longer allowed to perform the cast. The code that was here expresses this most directly (the recursive call does the "drilling down"), but as you note, it's inefficient in the case where we need to throw away the result of the recursive call.

I've changed this to use tail recursion and have added comments that hopefully make it clear what the recursion is doing.

ymand commented on clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp:
```diff
@@ -129,19 +207,19 @@ auto inPlaceClass() {
 auto isOptionalNulloptConstructor() {
   return cxxConstructExpr(
-      hasOptionalType(),
+      hasOptionalOrDerivedType(),
```
Here and below: Why admit a derived class here? I would think that we don't know (and therefore can't model) the definition of constructors, methods, operators on the derived class. I suppose I'm missing something?

martinboehme replied:
> I would think that we don't know (and therefore can't model) the definition of constructors, methods, operators on the derived class.

I agree in principle, but here we're looking specifically at a constructor that takes a `std::nullopt_t`. I'd say it's a pretty safe bet to assume this constructor is initializing the optional to be empty, even in a derived class (it would be _very_ confusing if it did anything else).

I believe it's also reasonable to conjecture about the behavior of the cases below that use `hasOptionalOrDerivedType()`, i.e. the `std::in_place_t` case and the "value or conversion constructor / assignment" cases.

`llvm::MaybeAlign` has [methods](https://github.com/llvm/llvm-project/blob/3e6db602918435b6a5ac476f63f8b259e7e73af4/llvm/include/llvm/Support/Alignment.h#L131) that correspond to these cases, and it seems desirable to model them correctly.

ymand reviewed (commented):
One high level question: is there a simpler path by which we avoid crashing on derived types but also don't support them? It seems like there is a lot of complexity required in getting this just right. Is it worth it (both in the code itself and in terms of runtime cost) to support these unusual cases? Of course, if it's necessary then the point is moot...

martinboehme commented:
> One high level question: is there a simpler path by which we avoid crashing on derived types but also don't support them? It seems like there is a lot of complexity required in getting this just right. Is it worth it (both in the code itself and in terms of runtime cost) to support these unusual cases? Of course, if it's necessary then the point is moot...

This is a good question. tl;dr: I think actually handling these cases instead of just ignoring them is the right tradeoff, but the situation certainly isn't clear cut.

It's _possible_ to avoid crashing on derived types but otherwise not support them. What we would need to do is call `RecordStorageLocation::getType()` on the storage location of the object that we're accessing; we would only proceed if we see an actual optional type (not a derived type). The thing is though that we would need to do this in _every_ `transfer_...` handler; this isn't ideal because it's easy to forget one, and then we get crashes there. I considered this option, but it seemed to me that the more principled and less error-prone approach is to actually treat derived types correctly.

As I say, the tradeoff isn't clear-cut. To summarize:

**Bail out but don't crash**
Pros:
*  Less conceptual complexity...
Cons:
*  ...but need a check in every handler (i.e. still lots of lines of code).
* If we forget to add the check everywhere, we get crashes.

**Correctly handle derived types**
Pros:
*  The additional code is centralized in one place.
*  Actually makes the check more expressive
Cons:
*  We need to walk the inheritance hierarchy and chains of casts, which adds complexity and potentially makes the check slower.

Will respond to the individual comments too, but probably not before Monday.

martinboehme committed:
fixup! [clang][dataflow] Make optional checker work for types derived from optional.

martinboehme commented:
Forgot to push a fixup commit -- now added.

ymand commented on clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp:
```diff
@@ -119,20 +119,28 @@ QualType getPublicType(const Expr *E) {
-    if (Base->getAccessSpecifier() != AS_public && !TyIsThisType)
+    if (Base->getAccessSpecifier() != AS_public && !CastingFromThis)
       break;
     Ty = Base->getType();
```
optional super nit: save `&Base` instead, avoiding a call to `getType()` until after the loop.

martinboehme replied:
Done.

More importantly than the minor efficiency gain, I think this makes the code more self-documenting. Thanks!

ymand commented on clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp:
```diff
@@ -129,19 +215,19 @@ auto inPlaceClass() {
 auto isOptionalNulloptConstructor() {
   return cxxConstructExpr(
-      hasOptionalType(),
+      hasOptionalOrDerivedType(),
```
Here and below -- now that this matcher is more expensive, please move until after the less expensive matchers.

martinboehme replied:
Done.

martinboehme committed:
fixup! fixup! [clang][dataflow] Make optional checker work for types derived from optional.
